### PR TITLE
Add an nginx docker image for use on the master.

### PR DESCRIPTION
**Pull request by brendandburns**
@ArtfulCoder @zmerlynn

### CODE REVIEW

File: cluster/images/nginx/nginx.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    "volumeMounts": [
+      { "name": "nginx",
+        "mountPath": "/etc/nginx",
+        "readOnly": false},
+      { "name": "k8s",
+        "mountPath": "/srv/kubernetes",
+        "readOnly": true},
+      { "name": "passwd",
+        "mountPath": "/usr/share/nginx",
+        "readOnly": false}
```
Comment (ArtfulCoder) [inline cluster/images/nginx/nginx.json]: do you need it to be writeable ?

File: cluster/images/nginx/nginx.json [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	],
+    "ports":[
+      { "name": "https",
+        "containerPort": 443,
+        "hostPort": 443}
+        ],
+    "volumeMounts": [
+      { "name": "nginx",
+        "mountPath": "/etc/nginx",
+        "readOnly": false},
```
Comment (ArtfulCoder) [inline cluster/images/nginx/nginx.json]: do you need it to be writeable ?

File: cluster/images/nginx/nginx.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+              "nginx",
+	      "-g",
+	      "daemon off;"
+	],
+    "ports":[
+      { "name": "https",
+        "containerPort": 443,
+        "hostPort": 443}
+        ],
+    "volumeMounts": [
```
Comment (ArtfulCoder) [inline cluster/images/nginx/nginx.json]: you might want to mount /var/log/nginx as well.
check out https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/saltbase/salt/nginx/nginx.conf#L38

File: cluster/images/nginx/README.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+### Nginx image
+This image is used to run nginx on the master.
+
+#### Instructions
+docker build -t gcr.io/google-containers/nginx:<version>
```
Comment (ArtfulCoder) [inline cluster/images/nginx/README.md]: I am curious why we didnt want to use the official nginx docker container https://registry.hub.docker.com/_/nginx/tags/

Comment (roberthbailey) [inline cluster/images/nginx/README.md]: This should be google_containers to match our other image paths (to support older versions of docker)

Comment (brendandburns) [inline cluster/images/nginx/README.md]: @ArtfulCoder there are complexities around licensing and derived images I'd rather avoid.

@roberthbailey done.

File: cluster/images/nginx/nginx.json (hunk 3 above)
Comment (brendandburns) [inline cluster/images/nginx/nginx.json]: done.

File: cluster/images/nginx/nginx.json (hunk 2 above)
Comment (brendandburns) [inline cluster/images/nginx/nginx.json]: fixed.

File: cluster/images/nginx/nginx.json (hunk 1 above)
Comment (brendandburns) [inline cluster/images/nginx/nginx.json]: fixed.

**Comment by brendandburns** [comment]
Comments addressed and PR updated.  ptal.

**Comment by ArtfulCoder** [comment]
LGTM
can be merged during business hours.

**Comment by brendandburns** [comment]
updated, to actually activate the container via Salt.

Please wait to merge tomorrow, I want to run e2e.  Will ping this PR when I have a passing e2e.

File: cluster/saltbase/salt/nginx/init.sls [hunk 5]
```diff
@@ -1,6 +1,8 @@
 nginx:
+{% if grains.cloud in ['gce'] %}
```
Comment (ArtfulCoder) [inline cluster/saltbase/salt/nginx/init.sls]: If gce, then nginx should not be installed isn't it ?

File: cluster/saltbase/salt/nginx/init.sls [hunk 6]
```diff
@@ -28,10 +30,20 @@ nginx:
...
+/etc/kubernetes/manifests/nginx.json:
+  file:
+    - managed
+    - source: salt://nginx/nginx.json
+    - user: root
+    - group: root
+    - mode: 644
+{% else %}
 nginx-service:
```
Comment (ArtfulCoder) [inline cluster/saltbase/salt/nginx/init.sls]: for gce, the service should be disabled.

File: cluster/saltbase/salt/nginx/init.sls [hunk 7]
```diff
@@ -28,10 +30,20 @@ nginx:
...
   service:
     - running
     - name: nginx
+{% endif %}
     - watch:
```
Comment (ArtfulCoder) [inline cluster/saltbase/salt/nginx/init.sls]: nginx will not be restarted anymore by kubelet if the watched file changes.
We will need to add logic to do it.

**Comment by ArtfulCoder** [comment]
Are you introducing this for gce only as a rollout strategy ? or is this change only for gce ?

File: cluster/saltbase/salt/nginx/init.sls (hunk 5 above)
Comment (brendandburns) [inline cluster/saltbase/salt/nginx/init.sls]: oops, yes.  fixed.

File: cluster/saltbase/salt/nginx/init.sls (hunk 6 above)
Comment (brendandburns) [inline cluster/saltbase/salt/nginx/init.sls]: done.

File: cluster/saltbase/salt/nginx/init.sls (hunk 7 above)
Comment (brendandburns) [inline cluster/saltbase/salt/nginx/init.sls]: actually, the kubelet does pick up changes to the file.  (although I agree that these particular watches don't do anything, moved into the else block)

Added "require" to ensure the sequence we need.

**Comment by brendandburns** [comment]
yes, only gce is a rollout strategy.  I'll move to AWS and others next.

**Comment by brendandburns** [comment]
Ok, this now passes e2e.  I actually added back in the install of the nginx .deb since it provides much of the /etc/nginx/... (e.g. mime types) files that we rely on.  We should really package all of this except the kubernetes site in the container image, but deferring that to a different PR (or when we eliminate nginx)

**Comment by brendandburns** [comment]
ptal.

Thanks!
--brendan

**Comment by dchen1107** [comment]
LGTM except I wish we have a Makefile introduced, so that we could track the versions of all bundled components including nginx here. 

To do this, you can simple convert the README file to a Makefile.

**Comment by dchen1107** [comment]
Also have you make sure cluster upgrade works too?

**Comment by ArtfulCoder** [comment]
@brendanburns You would need to disable service explicitly for gce so that nginx in a pod, can start successfully.

**Comment by brendandburns** [comment]
@dchen1107 
Makefile added.
Tested upgrade with:
```
git checkout master
make clean; make quick-release
cluster/kube-up.sh

git checkout nginx
make clean; make quick-release
cluster/kube-push.sh
```

And validate that the cluster came up ok in both cases.

@ArtfulCoder nginx disabled by salt.

ptal.
THanks
--brendan

**Comment by brendandburns** [comment]
Fixed after problems pointed out by @ArtfulCoder over IM.

**Comment by ArtfulCoder** [comment]
LGTM

**Comment by dchen1107** [comment]
LGTM

**Comment by dchen1107** [comment]
Great! one more step move forward for #5011.

**Comment by ArtfulCoder** [comment]
@brendanburns can we roll this out for other cloud providers too ?

**Comment by ArtfulCoder** [comment]
#6618

never mind.
 we are getting rid of nginx

**Pull request merged**